File: PAT_Basic_Level_by_go/1079 延迟的回文数/1079 延迟的回文数.go
/*
 * 题目：1079 延迟的回文数
 * 解题思路：
 * 给定一个正整数 A，将其逆转得 B，计算 A + B 的和 C。
 * 若 C 是回文数则输出，否则用 C 替换 A 继续上述过程，
 * 最多迭代 10 步。若 10 步内未出现回文数则输出失败信息。
 * 涉及大数加法，用字符串处理。
 */
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// reverse 返回字符串 s 的逆序字符串。
func reverse(s string) string {
	b := []byte(s)
	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
	return string(b)
}

// isPalindrome 判断字符串 s 是否为回文（正反相同）。
func isPalindrome(s string) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}

// addBig 实现两个长度相同的数字字符串的大数加法，返回结果字符串。
func addBig(a, b string) string {
	n := len(a)
	res := make([]byte, n)
	carry := 0
	// 从最低位开始逐位相加
	for i := n - 1; i >= 0; i-- {
		x := int(a[i] - '0')
		y := int(b[i] - '0')
		s := x + y + carry
		res[i] = byte('0' + s%10)
		carry = s / 10
	}
	// 最高位有进位则追加
	if carry > 0 {
		res = append([]byte{byte('0' + carry)}, res...)
	}
	// 去掉前导零
	out := strings.TrimLeft(string(res), "0")
	if out == "" {
		out = "0"
	}
	return out
}

// main 是程序入口函数，执行回文数迭代流程。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	if !scanner.Scan() {
		return
	}
	A := strings.TrimSpace(scanner.Text())
	// 若初始就是回文数，直接输出
	if isPalindrome(A) {
		fmt.Printf("%s is a palindromic number.\n", A)
		return
	}
	// 最多迭代 10 步
	for i := 0; i < 10; i++ {
		B := reverse(A)
		C := addBig(A, B)
		fmt.Printf("%s + %s = %s\n", A, B, C)
		if isPalindrome(C) {
			fmt.Printf("%s is a palindromic number.\n", C)
			return
		}
		A = C // 继续下一步迭代
	}
	fmt.Println("Not found in 10 iterations.")
}

File: PAT_Basic_Level_by_cpp/1079 延迟的回文数/1079 延迟的回文数.cpp
/*
 * 1079 延迟的回文数
 *
 * 【实现原理】
 * 本题涉及"回文数"和"大整数加法"操作。给定一个正整数（可能很大，超过普通整数范围），
 * 若其本身是回文数则直接输出；否则反复执行 N = N + reverse(N) 操作，
 * 最多迭代 10 次。若某次结果变为回文数则输出过程并结束；否则报告未找到。
 *
 * 【算法思路】
 * 1. 判断输入字符串 n 是否为回文（isPal 函数），若是则直接输出。
 * 2. 否则循环最多 10 次：
 *    a. 将 n 反转得到 r。
 *    b. 调用大整数加法 add(n, r) 计算和。
 *    c. 输出计算过程 "N + reverse(N) = sum"。
 *    d. 判断 sum 是否为回文，若是则输出结果并结束。
 *    e. 否则将 sum 赋给 n，继续下一轮迭代。
 * 3. 若 10 次后仍未找到回文，输出 "Not found in 10 iterations."。
 *
 * add 函数实现大整数加法：
 * - 从末位向首位逐位相加，处理进位。
 * - 最后若仍有进位则在最高位补 1。
 * - 将结果反转后返回（因为是从低位到高位构建的）。
 *
 * 【复杂度分析】
 * - 时间复杂度：O(k·L)，其中 k 为迭代次数（≤10），L 为数字字符串长度。
 * - 空间复杂度：O(L)，用于存储大整数结果。
 */
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

/*
 * 大整数加法：将两个等长的数字字符串相加（a 和 b 长度相同）
 * 从最低位（末尾）向最高位逐位相加，处理进位
 */
string add(const string& a, const string& b) {
    string res;
    int carry = 0;                             // 进位标志
    for (int i = (int)a.size() - 1; i >= 0; --i) {
        int s = (a[i] - '0') + (b[i] - '0') + carry;
        res += (char)('0' + s % 10);           // 当前位数字
        carry = s / 10;                        // 更新进位
    }
    if (carry) res += (char)('0' + carry);     // 最高位仍有进位则补 1
    reverse(res.begin(), res.end());           // 反转得到正确顺序
    return res;
}

// 判断字符串是否为回文
bool isPal(const string& s) {
    string t = s;
    reverse(t.begin(), t.end());
    return t == s;
}

int main() {
    string n;
    if (!(cin >> n)) return 0;

    // 若初始即为回文数，直接输出
    if (isPal(n)) {
        cout << n << " is a palindromic number." << endl;
        return 0;
    }

    // 最多迭代 10 次
    for (int i = 0; i < 10; ++i) {
        string r = n;
        reverse(r.begin(), r.end());           // 反转得到 reverse(N)
        string sum = add(n, r);                // 大整数加法
        cout << n << " + " << r << " = " << sum << endl;
        if (isPal(sum)) {
            cout << sum << " is a palindromic number." << endl;
            return 0;
        }
        n = sum;                               // 更新 N 为新的和，继续迭代
    }
    cout << "Not found in 10 iterations." << endl;
    return 0;
}
